### PR TITLE
dolibarr: init at 15.0.3, nixos/dolibarr: init

### PR DESCRIPTION
**Pull request by RaitoBezarius**
###### Description of changes

Taken from @aanderse branch, this is Dolibarr packaging & NixOS module.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
@aanderse If you do have some time, I would be interested into knowing if you could install Dolibarr and the `conf.php`, I'm getting errors (the NixOS test reproduce it) of the following form:
```
<!DOCTYPE HTML>
<html>
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Dolibarr installer">
<link rel="stylesheet" type="text/css" href="default.css">
<!-- Includes CSS for JQuery -->
<link rel="stylesheet" type="text/css" href="../includes/jquery/css/base/jquery-ui.min.css" />
<!-- Includes JS for JQuery -->
<script type="text/javascript" src="../includes/jquery/js/jquery.min.js"></script>
<script type="text/javascript" src="../includes/jquery/js/jquery-ui.min.js"></script>
<title>Dolibarr install or upgrade</title>
</head>
<body>
<div class="divlogoinstall" style="text-align:center"><img class="imglogoinstall" src="../theme/dolibarr_logo.svg" alt="Dolibarr logo" width="300px"><br>15.0.3</div><br><span class="titre">Dolibarr install or upgrade - Configuration file</span>
<form name="forminstall" style="width: 100%" action="step1.php" method="POST">
<input type="hidden" name="testpost" value="ok">
<input type="hidden" name="action" value="set">
<table class="main" width="100%"><tr><td>
<table class="main-inside-bis" width="100%"><tr><td>
Configuration file <b>/var/lib/dolibarr/conf.php</b> is not writable. Check permissions. For first install, your web server must be able to write into this file during configuration process ("chmod 666" for example on a Unix like OS).</td></tr></table>
</td></tr></table>
<!-- pFooter -->
<input type="hidden" name="selectlang" value="auto"></form>

<!-- Start of log output
2022-09-03 22:37:45 DEBUG - fileconf: entering fileconf.php page<br>
2022-09-03 22:37:45 WARN fileconf: config file is not writable<br>
2022-09-03 22:37:45 DEBUG - fileconf: end<br>
End of log output -->

</body>
</html>
```

My understanding though is that php-fpm is supposed to have the rights to edit the state directory, therefore, creating the configuration file, but I have no clue about what is going on, is it inheriting the rights from NGINX or something?

File: nixos/modules/services/web-apps/dolibarr.nix [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+    force_install_nophpinfo = true;
+    force_install_lockinstall = "444";
+    force_install_distrib = "nixos";
+    force_install_type = "mysqli";
+    force_install_dbserver = cfg.database.host;
+    force_install_port = toString cfg.database.port;
+    force_install_database = cfg.database.name;
+    force_install_databaselogin = cfg.database.user;
+
+    force_install_mainforcehttps = false; # TODO: true
```
Comment (aanderse) [inline nixos/modules/services/web-apps/dolibarr.nix:40]: Infer from `nginx` configuration?

File: nixos/modules/services/web-apps/dolibarr.nix [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+      ;
+    in
+      pkgs.writeText filename ''
+        <?php
+        ${concatStringsSep "\n" (mapAttrsToList (k: v: "\$${k} = ${toStr k v};") settings)}
+      '';
+
+  # see https://github.com/Dolibarr/dolibarr/blob/develop/htdocs/install/install.forced.sample.php for all possible values
+  install = {
+    force_install_noedit = 2; # was 1;
```
Comment (aanderse) [inline nixos/modules/services/web-apps/dolibarr.nix:29]: Sorry, can drop this comment.

**Review by aanderse — CHANGES_REQUESTED** [review]
Looking good. Big fan of the test.

> @aanderse If you do have some time

I'll try to find some time 👍

File: pkgs/servers/web-apps/dolibarr/default.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      --replace '//$conffile = ' '$conffile = ' \
+      --replace '//$conffiletoshow = ' '$conffiletoshow = '
+  '';
+
+  installPhase = ''
+    mkdir -p "$out"
+    cp -r * $out
+  '';
+
+  meta = with lib; {
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply pkgs/servers/web-apps/dolibarr/default.nix:33]: Yes, will add myself.

File: nixos/modules/services/web-apps/dolibarr.nix [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    ]);
+
+    services.phpfpm.pools.dolibarr = {
+      inherit (cfg) user group;
+      phpPackage = pkgs.php.buildEnv {
+        # recommended by dolibarr web application
+        extraConfig = ''
+          session.use_strict_mode = 1
+          session.cookie_samesite = "Lax"
+          ; open_basedir = "${package}/htdocs, ${cfg.stateDir}"
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply nixos/modules/services/web-apps/dolibarr.nix:265]: Indeed, my understanding is that it behaves quite bad with symlinks, do we have reasons to believe we can get it to work at all?

File: pkgs/servers/web-apps/dolibarr/default.nix [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  src = fetchFromGitHub {
+    owner = "Dolibarr";
+    repo = "dolibarr";
+    rev = version;
+    sha256 = "sha256-HMOYj93ZvqM0FQjt313yuGj/r9ELqQlnNkg/CxrBjRM=";
+  };
+
+  dontBuild = true;
+
+  postPatch = ''
```
[the remark this answers is not in the document]
Comment (RaitoBezarius) [reply pkgs/servers/web-apps/dolibarr/default.nix:16]: Heh :-)
Seems like the machinery for reseating all the paths with the state directory parameter, but I'm not certain for the last two substituteInPlace.

File: nixos/modules/services/web-apps/dolibarr.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (aanderse) [reply nixos/modules/services/web-apps/dolibarr.nix:265]: Ah OK makes sense. Thanks!

**Comment by aanderse** [comment]
[This](https://github.com/NixOS/nixpkgs/commit/5616c89c58db1ddf33659fbd564a2690a5e6ae5e) should do it. `dolibarr-config.service` should have always been a `oneshot`.

**Comment by RaitoBezarius** [comment]
> [This](https://github.com/NixOS/nixpkgs/commit/5616c89c58db1ddf33659fbd564a2690a5e6ae5e) should do it.
> 
> * apparently the installer requires that `conf.php` exists, even if empty... man_shrugging
> * `dolibarr-config.service` should have always been a `oneshot`
> * we don't need to run `dolibarr-config.service` explicitly, though it might be a good idea to run it once more at the end of this test

Trying it now.

**Comment by RaitoBezarius** [comment]
Nice, it is writable, and I'm running into stuff I can grok at:
```
machine #   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0[   11.382551] nginx[921]: 2022/09/04 00:01:44 [error] 921#921: *8 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught mysqli_sql_exception: No such file or directory in /nix/store/xy62178cl1ynk86qf2if97hcpr1nlygj-dolibarr-15.0.3/htdocs/core/db/mysqli.class.php:218
machine # 100  1218    0  1179  100    39   7315    241 --:--:-- --:--:-[   11.391545] nginx[921]: Stack trace:
machine # [   11.392370] nginx[921]: #0 /nix/store/xy62178cl1ynk86qf2if97hcpr1nlygj-dolibarr-15.0.3/htdocs/core/db/mysqli.class.php(218): mysqli->__construct()
machine # - --:--:--  7612
machine # [   11.398301] nginx[921]: #1 /nix/store/xy62178cl1ynk86qf2if97hcpr1nlygj-dolibarr-15.0.3/htdocs/core/db/mysqli.class.php(93): DoliDBMysqli->connect()
machine # [   11.410389] nginx[921]: #2 /nix/store/xy62178cl1ynk86qf2if97hcpr1nlygj-dolibarr-15.0.3/htdocs/core/lib/functions.lib.php(86): DoliDBMysqli->__construct()
```

File: nixos/modules/services/web-apps/dolibarr.nix [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        type = types.bool;
+        default = true;
+        description = "Create the database and database user locally.";
+      };
+    };
+
+    settings = mkOption {
+      type = with types; (attrsOf (oneOf [ bool int str ]));
+      default = { };
+      description = lib.mdDoc "Dolibarr settings, see <> for details.";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/dolibarr.nix:134]: Any details here?

File: pkgs/servers/web-apps/dolibarr/default.nix [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    mkdir -p "$out"
+    cp -r * $out
+  '';
+
+  passthru.tests = { inherit (nixosTests) dolibarr; };
+
+  meta = with lib; {
+    description = "A enterprise resource planning (ERP) and customer relationship manager (CRM) server";
+    homepage = "https://dolibarr.org/";
+    license = licenses.gpl3;
```
Comment (aanderse) [inline pkgs/servers/web-apps/dolibarr/default.nix:38]: `gpl3Plus` or ...`gpl3Only`?

**Review by aanderse — CHANGES_REQUESTED** [review]
Looks good. Just a few tiny this and time to merge 🎉

File: pkgs/servers/web-apps/dolibarr/default.nix (hunk 7 above)
Comment (RaitoBezarius) [reply pkgs/servers/web-apps/dolibarr/default.nix]: Well, they say both is fine:
> The Dolibarr software as a whole is distributed under the GNU General Public License as published by the Free Software Foundation;
either version 3 of the License, or (at your option) any later version (GPL-3+). More information: https://www.gnu.org/licenses/gpl-3.0.txt.
> A copy of this license is available in the COPYING file.

I have changed to GPL3+.

File: nixos/modules/services/web-apps/dolibarr.nix (hunk 6 above)
Comment (RaitoBezarius) [reply nixos/modules/services/web-apps/dolibarr.nix]: Fixed

**Pull request merged**